packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts

```diff
@@ -36,28 +54,139 @@ export default createRule<Options, MessageIds>({
+            type: 'boolean',
+          },
           requireDefaultForNonUnion: {
             description: `If 'true', require a 'default' clause for switches on non-union types.`,
             type: 'boolean',
           },
         },
+        additionalProperties: false,
       },
     ],
     messages: {
       switchIsNotExhaustive:
         'Switch is not exhaustive. Cases not matched: {{missingBranches}}',
+      dangerousDefaultCase:
+        'The switch statement is exhaustive, so the default case is unnecessary.',
       addMissingCases: 'Add branches for missing cases.',
     },
   },
-  defaultOptions: [{ requireDefaultForNonUnion: false }],
-  create(context, [{ requireDefaultForNonUnion }]) {
+  defaultOptions: [
+    {
+      allowDefaultCaseForExhaustiveSwitch: true,
+      requireDefaultForNonUnion: false,
+    },
+  ],
+  create(
+    context,
+    [{ allowDefaultCaseForExhaustiveSwitch, requireDefaultForNonUnion }],
+  ) {
     const sourceCode = getSourceCode(context);
     const services = getParserServices(context);
     const checker = services.program.getTypeChecker();
     const compilerOptions = services.program.getCompilerOptions();
 
+    /**
+     * @returns Metadata about whether the switch is exhaustive (or `undefined`
+     *          if the switch case is not a union).
+     */
+    function getSwitchStatementMetadata(
+      node: TSESTree.SwitchStatement,
+    ): SwitchStatementMetadata {
+      const defaultCase = node.cases.find(
+        switchCase => switchCase.test == null,
+      );
+
+      const discriminantType = getConstrainedTypeAtLocation(
+        services,
+        node.discriminant,
+      );
+
+      const symbolName = discriminantType.getSymbol()?.escapedName as
+        | string
+        | undefined;
+
+      if (!discriminantType.isUnion()) {
+        return {
+          symbolName,
+          missingBranchTypes: [],
+          defaultCase,
+          isUnion: false,
+        };
+      }
+
+      const caseTypes = new Set<ts.Type>();
+      for (const switchCase of node.cases) {
+        // If the `test` property of the switch case is `null`, then we are on a
+        // `default` case.
+        if (switchCase.test == null) {
+          continue;
+        }
+
+        const caseType = getConstrainedTypeAtLocation(
+          services,
+          switchCase.test,
+        );
+        caseTypes.add(caseType);
+      }
+
+      const unionTypes = tsutils.unionTypeParts(discriminantType);
+      const missingBranchTypes = unionTypes.filter(
+        unionType => !caseTypes.has(unionType),
+      );
+
+      return {
+        symbolName,
+        missingBranchTypes,
+        defaultCase,
+        isUnion: true,
+      };
+    }
+
+    function checkSwitchExhaustive(
+      node: TSESTree.SwitchStatement,
+      switchStatementMetadata: SwitchStatementMetadata,
+    ): void {
+      const { missingBranchTypes, symbolName, defaultCase } =
+        switchStatementMetadata;
```

Comment:
[Style] Is the reason for not destructuring inline so as to not make them break across lines? If so, have you considered changing the interface name to `SwitchMetadata`? The _"Statement"_ in there is arguably inferable from the context anyway...

```suggestion
      { missingBranchTypes, symbolName, defaultCase }: SwitchMetadata,
    ): void {
```

This is very much not a blocker at all 😄 just mentioning in case you like it.

Reply:
i renamed it to switchMetadata now, thanks.

as for why I destructured it, it was for two reasons:
1) to avoid repetition, i.e. `if (x + y + z)` is easier to read than `if (switchMetadata.x + switchMetadata.y + switchMetadata.z)`
2) to avoid prettier breaking lines

Reply:
i guess another benefit is that it prevents mutation, since we are working on a non-read-only version of the object, and thus we should make copies of the values instead of working with them on the object directly, if we did that they we would be more likely to accidently mutate it